lib/Monads/trace/Trace_Monad_Equations.thy

```diff
@@ -10,41 +10,364 @@
+  apply (cases P)
+   apply (simp add: whenE_def)
+  apply simp
+  done
+
+lemma maybe_fail_bind_fail:
+  "unless P fail >>= (\<lambda>_. fail) = fail"
+  "when P fail >>= (\<lambda>_. fail) = fail"
+  by (clarsimp simp: bind_def fail_def return_def
+                     unless_def when_def)+
+
+lemma select_singleton[simp]:
   "select {x} = return x"
   by (simp add: select_def return_def)
 
+lemma return_modify:
+  "return () = modify id"
+  by (simp add: return_def simpler_modify_def)
+
+lemma liftE_liftM_liftME:
+  "liftE (liftM f m) = liftME f (liftE m)"
+  by (simp add: liftE_liftM liftME_liftM liftM_def)
+
+lemma bind_return_unit:
+  "f = (f >>= (\<lambda>x. return ()))"
+  by simp
+
+lemma modify_id_return:
+ "modify id = return ()"
+  by (simp add: simpler_modify_def return_def)
+
+
+lemma liftE_bind_return_bindE_returnOk:
+  "liftE (v >>= (\<lambda>rv. return (f rv)))
+     = (liftE v >>=E (\<lambda>rv. returnOk (f rv)))"
+  by (simp add: liftE_bindE, simp add: liftE_def returnOk_def)
+
+lemma bind_eqI:
+  "g = g' \<Longrightarrow> f >>= g = f >>= g'" by simp
+
+lemma unlessE_throwError_returnOk:
+  "(if P then returnOk v else throwError x)
+    = (unlessE P (throwError x) >>=E (\<lambda>_. returnOk v))"
+  by (cases P, simp_all add: unlessE_def)
+
+lemma gets_the_bind_eq:
+  "\<lbrakk> f s = Some x; g x s = h s \<rbrakk>
+    \<Longrightarrow> (gets_the f >>= g) s = h s"
+  by (simp add: gets_the_def bind_assoc exec_gets assert_opt_def)
+
+lemma zipWithM_x_modify:
+  "zipWithM_x (\<lambda>a b. modify (f a b)) as bs
+   = modify (\<lambda>s. foldl (\<lambda>s (a, b). f a b s) s (zip as bs))"
+  apply (simp add: zipWithM_x_def zipWith_def sequence_x_def)
+  apply (induct ("zip as bs"))
+   apply (simp add: simpler_modify_def return_def)
+  apply (rule ext)
+  apply (simp add: simpler_modify_def bind_def split_def)
+  done
+
+lemma assert2:
+  "(do v1 \<leftarrow> assert P; v2 \<leftarrow> assert Q; c od)
+     = (do v \<leftarrow> assert (P \<and> Q); c od)"
+  by (simp add: assert_def split: if_split)
+
+lemma assert_opt_def2:
+  "assert_opt v = (do assert (v \<noteq> None); return (the v) od)"
+  by (simp add: assert_opt_def split: option.split)
+
+lemma gets_assert:
+  "(do v1 \<leftarrow> assert v; v2 \<leftarrow> gets f; c v1 v2 od)
+     = (do v2 \<leftarrow> gets f; v1 \<leftarrow> assert v; c v1 v2 od)"
+  by (simp add: simpler_gets_def return_def assert_def fail_def bind_def
+         split: if_split)
+
+lemma modify_assert:
+  "(do v2 \<leftarrow> modify f; v1 \<leftarrow> assert v; c v1 od)
+    = (do v1 \<leftarrow> assert v; v2 \<leftarrow> modify f; c v1 od)"
+  by (simp add: simpler_modify_def return_def assert_def fail_def bind_def
+         split: if_split)
+
+lemma gets_fold_into_modify:
+  "do x \<leftarrow> gets f; modify (g x) od = modify (\<lambda>s. g (f s) s)"
+  "do x \<leftarrow> gets f; _ \<leftarrow> modify (g x); h od
+     = do modify (\<lambda>s. g (f s) s); h od"
```

Comment:
there's a bunch of wonky indentation in this file, but I'm guessing these are moved/copied?

Reply:
Yeah, they were copied from nondet. Now that you've pointed them out I've taken the opportunity to fix the indentation up, both here and in the original nondet files. It's almost definitely still not perfect but should at least be an improvement.